plasmapy/utils/decorators/checks.py

```diff
@@ -1,20 +1,1219 @@
+    #
+    __check_defaults = {
+        'units': None,
+        'equivalencies': None,
+        'pass_equivalent_units': False,
+        'none_shall_pass': False,
+    }
+
+    def __init__(
+            self,
+            checks_on_return: Union[u.Unit,
+                                    List[u.Unit],
+                                    Dict[str, Any]] = None,
+            **checks: Union[u.Unit, List[u.Unit], Dict[str, Any]]):
+
+        super().__init__(checks_on_return=checks_on_return, **checks)
+
+    def __call__(self, f):
+        """
+        Parameters
+        ----------
+        f
+            Function to be wrapped
+
+        Returns
+        -------
+        function
+            wrapped function of `f`
+        """
+        self.f = f
+        wrapped_sign = inspect.signature(f)
+
+        @preserve_signature
+        @functools.wraps(f)
+        def wrapper(*args, **kwargs):
+            # combine args and kwargs into dictionary
+            bound_args = wrapped_sign.bind(*args, **kwargs)
+            bound_args.apply_defaults()
+
+            # get checks
+            checks = self._get_unit_checks(bound_args)
+
+            # check (input) argument units
+            for arg_name in checks:
+                # skip check of output/return
+                if arg_name == 'checks_on_return':
+                    continue
+
+                # check argument
+                self._check_unit(bound_args.arguments[arg_name],
+                                 arg_name,
+                                 checks[arg_name])
+
+            # call function
+            _return = f(**bound_args.arguments)
+
+            # check output
+            if 'checks_on_return' in checks:
+                self._check_unit(_return, 'checks_on_return',
+                                 checks['checks_on_return'])
+
+            return _return
+        return wrapper
+
+    def _get_unit_checks(
+            self,
+            bound_args: inspect.BoundArguments
+    ) -> Dict[str, Dict[str, Any]]:
+        """
+        Review :attr:`checks` and function bound arguments to build a complete 'checks'
+        dictionary.  If a check key is omitted from the argument checks, then a default
+        value is assumed (see `check units`_)
+
+        Parameters
+        ----------
+        bound_args: :class:`inspect.BoundArguments`
+            arguments passed into the function being wrapped
+
+            .. code-block:: python
+
+                bound_args = inspect.signature(f).bind(*args, **kwargs)
+
+        Returns
+        -------
+        Dict[str, Dict[str, Any]]
+            A complete 'checks' dictionary for checking function input arguments
+            and return.
+        """
+        # initialize validation dictionary
+        out_checks = {}
+
+        # Iterate through function bound arguments + return and build `out_checks`:
+        #
+        # artificially add "return" to parameters
```

Comment:
I do it this way because we scan through all `signature(foo).parameters` to review annotations and compare them to **checks**.  However, `signature(foo)` does not include the return annotations in the `parameters` dictionary.   I don't want to write separate code to handle return checks, so I artificially create a `'checks_on_return'` parameter (`inspect.Parameter()`) with the return annotations.  This way, the input parameters and return parameters run through the same "checks" code.